File: 11-container-with-most-water/11-container-with-most-water.py
class Solution:
    
    def maxArea(self, height: List[int]) -> int:
      def area(l, r) -> int:
        return (r-l)*min(height[l], height[r])
      l = 0
      r = len(height) - 1
      result = 0
      lastH = 0
      left = height[l]
      right = height[r]
      while (l < r):
        minH = min(left, right)
        if (minH > lastH):
          result = max(result, minH * (r - l))
          lastH = minH
        
        if left < right: 
          l += 1
          left = height[l]
        elif (left == right):
          l += 1
          left = height[l]
          r -= 1
          right = height[r]
        else: 
          r -= 1
          right = height[r]
        
      return result
        

File: 11-container-with-most-water/11-container-with-most-water.go
func min(a, b int) int {
  if a > b {
    return b
  } else {
    return a
  }
}

func maxArea(height []int) int {
  l, r := 0, len(height) - 1
  h := min(height[l], height[r])
  ans := (r - l)  * h
  for l < r {
    for ; l < r && height[l] <= h; l++ {}
    for ; l < r && height[r] <= h; r-- {}
    if l >= r { break }
    h = min(height[l], height[r])
    area := (r-l) * h
    if area > ans {
      ans = area
    }
  }
  return ans
}